docker-sqitch.bat

lastlink opened the pull request:
* init windows script 
  * converted docker-sqitch.sh to docker-sqitch.bat for windows, runs the container and set variables w/ argument, working command binding
  * been following along the postgres tutorial and windows cmd working so far.
* fixed docker-sqitch-sh to work on git bash on a windows machine, must be called
  * `sh docker-sqitch.sh help`
* updated readme for windows and mac
* pipeline updated for separate one only for prs

fixes #31

theory commented:
Oh, amazing, have needed this for a while, thanks! LMK when it’s ready.

Should probably consider adding something to the test workflow to make sure this scripts work properly…

lastlink commented:
@theory the test workflow pipeline may need fixing, this file makes no changes, but the test flow is failing. It now successfully mounts, idk about getting the help command to work.

theory commented:
It looks like the Docker Hub auth fails. WTF.

lastlink commented:
@theory I've separated the pull request pipeline to only validate the build, the other one will still run once a month and on merges to **main**, I think the docker key may need to be fixed for auth.
Also I've updated the readme for using the sh script on a windows machine. The .bat file is progress in the right direction I'll see if I can get the mount working.

lastlink commented:
@theory .bat file for windows cmd is **READY**!!. Also I updated the readme for clearer instructions on running mac, windows. I also added instructions for using original script using git bash on windows.

## notes
.bat script will run the help command correctly and shuts off docker image after closing. I did not understand `# Iterate over optional Sqitch and engine variables.` I need an example of running the sh command w/ the optional variables in the bash shell to understand how to correct, I did a similar thing for the bat file that will check if the optional variable is set and add the argument. Also did not understand what `# Determine the name of the container home directory.` should do but I set the default regardless.

theory commented on README.md:
```diff
@@ -1,9 +1,32 @@
-Sqitch Docker Packaging
+# Sqitch Docker Packaging
```
Should not be needed with the `=====` line, but you can delete that line if the goal is consistent header markup.

lastlink replied:
reverted

theory commented on README.md:
````diff
@@ -1,9 +1,32 @@
+
+### linux,mac
+
+```sh
````
If you use the fencepost syntax, then the leading four spaces should be removed form the lines of code.

lastlink replied:
fixed

theory commented on README.md:
````diff
@@ -1,9 +1,32 @@
+```bat
+docker pull sqitch/sqitch
+curl -L https://raw.githubusercontent.com/sqitchers/docker-sqitch/main/docker-sqitch.bat -o sqitch.bat
+sqitch.bat help
````
Does it even need the `.bat`?

lastlink replied:
no, fixed

theory commented:
>  I did not understand `# Iterate over optional Sqitch and engine variables.`

It just looks to see if each variable is set, and if so, passes it to the container. That said, I *think* it may not be necessary to test them, just pass them all to docker, and it will set them if the exist and ignore them if they don't. Not sure if we could run into command-line length issues, though.

>  Also did not understand what `# Determine the name of the container home directory.` should do but I set the default regardless.

Inside the container, the home directory is `/home` unless you run it as root, in which case it/s `/root` inside the container. This is important for mapping the the home directory to the proper home directory inside the container, so that one's configuration files get read. I don't know how much this would impact Windows; does one use a file named `.pgpass` in one's home directory on Windows? [Doesn't look like it](https://www.postgresql.org/docs/current/libpq-pgpass.html). So to be fully transparent will be more difficult on Windows, since ` %APPDATA%\postgresql\pgpass.conf` will not be converted to `~/.pgass` inside the container. That issue will likely hold for a bunch of other database clients too, I expect.

theory commented on README.md:
````diff
@@ -1,9 +1,32 @@
+```sh
+docker pull sqitch/sqitch
+curl -L https://git.io/JJKCn -o sqitch
+sh sqitch help
````
Can't you also `chmod +x sqitch` on git bash and have it work without passing it to `sh`?

lastlink replied:
this was from earlier testing w/ cmd, chmod does work in git bash, so I've removed the separate documentation for it.

theory commented:
Merged, thank you! Quick note: I deleted the ` MINGW32_NT*|MINGW64_NT*)` bit because `$SQITCH_ORIG_FULLNAME` is supposed to be a full name, not a username. LMK if I'm making a false assumption here.

Also, should `echo %passopt%` be commented out?

theory commented:
Oh, and is `curl` standard on Windows boxes now?

lastlink commented:
Curl works on windows, however, not all the parameters work. Iv had to tweak added parameters on it before. Also for the full username without a tweak on Windows it's like a 50 line username info so I was giving it the short name. It needs to be filtered down on windows.

lastlink commented:
It's wget that doesn't work on windows. I don't know when curl became standard on Windows, I think it also works on windows 7.

theory commented:
> Also for the full username without a tweak on Windows it's like a 50 line username info so I was giving it the short name. It needs to be filtered down on windows.

Something [like this](https://stackoverflow.com/a/34647284) should do the trick, eh? I tried on my work PC, but, oddly, my username doesn't seem to be included in `net user`. Does it work for you? (Change `%USERNAME%` to `$USERNAME` for git-bash.
